File: src/focus_agent/api/contract_models/agent.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class AgentRolePolicyResponse(BaseModel):
    enabled: bool = False
    default_model: str
    helper_model: str | None = None
    max_parallel_runs: int = 1
    roles: list[str] = Field(default_factory=list)
    role_models: dict[str, str | None] = Field(default_factory=dict)
    fallback_order: list[str] = Field(default_factory=list)


class AgentRoleDryRunRequest(BaseModel):
    message: str
    scene: str = "long_dialog_research"
    skill_hints: list[str] = Field(default_factory=list)
    available_tools: list[str] = Field(default_factory=list)


class AgentRoleDryRunResponse(BaseModel):
    policy: AgentRolePolicyResponse
    plan: dict[str, Any] = Field(default_factory=dict)


class AgentSkillSelectRequest(BaseModel):
    message: str
    skill_hints: list[str] = Field(default_factory=list)
    semantic_enabled: bool | None = None
    semantic_threshold: float | None = None


class AgentSkillSemanticCandidateResponse(BaseModel):
    skill_id: str
    score: float
    matched_terms: list[str] = Field(default_factory=list)
    auto_activate: bool = False
    rationale: str = ""


class AgentSkillSelectionResponse(BaseModel):
    selection_id: str | None = None
    skill_ids: list[str] = Field(default_factory=list)
    stripped_message: str = ""
    prompt_mode: str | None = None
    selection_source: str = "none"
    matched_triggers: list[str] = Field(default_factory=list)
    semantic_candidates: list[AgentSkillSemanticCandidateResponse] = Field(default_factory=list)
    confidence: float = 0.0
    rationale: str = ""
    semantic_enabled: bool = True
    semantic_threshold: float = 0.22


class AgentSkillSelectionEventResponse(BaseModel):
    selection_id: str
    user_id: str | None = None
    message_preview: str | None = None
    selection_source: str = "none"
    explicit_hints: list[str] = Field(default_factory=list)
    activated_skill_ids: list[str] = Field(default_factory=list)
    matched_triggers: list[str] = Field(default_factory=list)
    semantic_candidates: list[dict[str, Any]] = Field(default_factory=list)
    confidence: float = 0.0
    rationale: str = ""
    feedback: str | None = None
    feedback_reason: str | None = None
    user_override: dict[str, Any] = Field(default_factory=dict)
    created_at: str | None = None
    updated_at: str | None = None


class AgentSkillSelectionEventListResponse(BaseModel):
    items: list[AgentSkillSelectionEventResponse] = Field(default_factory=list)
    count: int = 0
    filters: dict[str, Any] = Field(default_factory=dict)
    limit: int = 50
    backend: str = "governance"
    available: bool = True


class AgentSkillSelectionFeedbackRequest(BaseModel):
    feedback: str
    reason: str | None = None
    user_override: dict[str, Any] = Field(default_factory=dict)


class AgentSkillSelectionFeedbackResponse(BaseModel):
    item: AgentSkillSelectionEventResponse
    feedback_event_id: str | None = None


class AgentFeedbackTrendResponse(BaseModel):
    negative_feedback_count: int = 0
    merge_review_apply_success_rate: float | None = None
    merge_review_conflict_rate: float | None = None
    skill_low_confidence_rate: float | None = None
    skill_override_rate: float | None = None
    context_high_drift_count: int = 0
    notes_tasks_capture_count: int = 0
    top_failing_trajectory_samples: list[dict[str, Any]] = Field(default_factory=list)
    generated_at: str | None = None


class AgentSkillPreferenceRequest(BaseModel):
    state: str = "default"
    metadata: dict[str, Any] = Field(default_factory=dict)


class AgentSkillPreferenceResponse(BaseModel):
    preference_id: str
    user_id: str
    skill_id: str
    state: str = "default"
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: str | None = None
    updated_at: str | None = None


class AgentSkillCatalogItemResponse(BaseModel):
    skill_id: str
    description: str = ""
    triggers: list[str] = Field(default_factory=list)
    when_to_use: list[str] = Field(default_factory=list)
    recommended_tools: list[str] = Field(default_factory=list)
    prompt_mode: str | None = None
    path: str | None = None
    preference: AgentSkillPreferenceResponse | None = None


class AgentSkillCatalogResponse(BaseModel):
    items: list[AgentSkillCatalogItemResponse] = Field(default_factory=list)
    count: int = 0


class AgentContextExplainRequest(BaseModel):
    thread_id: str | None = None
    turn_id: str | None = None
    selected_memories: list[dict[str, Any]] = Field(default_factory=list)
    excluded_memories: list[dict[str, Any]] = Field(default_factory=list)
    compaction_summary: dict[str, Any] = Field(default_factory=dict)
    drift_report: dict[str, Any] = Field(default_factory=dict)
    artifact_refs: list[dict[str, Any]] = Field(default_factory=list)
    token_counting: dict[str, Any] = Field(default_factory=dict)
    risk_flags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class AgentContextEvidenceResponse(BaseModel):
    evidence_id: str
    user_id: str | None = None
    thread_id: str | None = None
    turn_id: str | None = None
    source_kind: str = "context_explain"
    selected_memories: list[dict[str, Any]] = Field(default_factory=list)
    excluded_memories: list[dict[str, Any]] = Field(default_factory=list)
    compaction_summary: dict[str, Any] = Field(default_factory=dict)
    drift_report: dict[str, Any] = Field(default_factory=dict)
    artifact_refs: list[dict[str, Any]] = Field(default_factory=list)
    token_counting: dict[str, Any] = Field(default_factory=dict)
    risk_flags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: str | None = None


class AgentContextEvidenceListResponse(BaseModel):
    items: list[AgentContextEvidenceResponse] = Field(default_factory=list)
    count: int = 0
    filters: dict[str, Any] = Field(default_factory=dict)
    limit: int = 50
    backend: str = "governance"
    available: bool = True


class AgentContextExplainResponse(BaseModel):
    item: AgentContextEvidenceResponse


class AgentRoleDecisionListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentCapabilityResponse(BaseModel):
    name: str
    description: str = ""
    toolset: str | None = None
    allowed_roles: list[str] = Field(default_factory=list)
    risk_level: str = "low"
    side_effect: bool = False
    parallel_safe: bool = False
    cacheable: bool = False
    requires_network: bool = False
    requires_workspace_write: bool = False
    requires_approval: bool = False
    usage_examples: list[str] = Field(default_factory=list)
    negative_examples: list[str] = Field(default_factory=list)
    max_calls_per_turn: int | None = None
    output_summary_contract: str | None = None
    sensitive_args: list[str] = Field(default_factory=list)
    redaction_policy: str = "mask"
    provider_id: str | None = None


class AgentCapabilityListResponse(BaseModel):
    items: list[AgentCapabilityResponse] = Field(default_factory=list)
    count: int = 0


class AgentToolsetResponse(BaseModel):
    name: str
    description: str = ""
    tools: list[str] = Field(default_factory=list)
    count: int = 0
    provider_ids: list[str] = Field(default_factory=list)
    risk_levels: list[str] = Field(default_factory=list)
    allowed_roles: list[str] = Field(default_factory=list)
    intent_policies: list[str] = Field(default_factory=list)
    requires_network: bool = False
    requires_workspace_write: bool = False
    side_effect: bool = False
    requires_approval: bool = False


class AgentToolsetListResponse(BaseModel):
    items: list[AgentToolsetResponse] = Field(default_factory=list)
    count: int = 0


class AgentToolRouteRequest(BaseModel):
    role: str = "executor"
    tool_policy: str = "execution"
    available_tools: list[str] = Field(default_factory=list)
    enforce: bool | None = None


class AgentToolRouteResponse(BaseModel):
    plan: dict[str, Any] = Field(default_factory=dict)


class AgentToolRouteDecisionListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentMemoryCuratorPolicyResponse(BaseModel):
    enabled: bool = False
    auto_promote_on_merge: bool = True
    branch_local_only_until_merge: bool = True
    conflict_strategy: str = "needs_review"


class AgentMemoryCuratorEvaluateRequest(BaseModel):
    root_thread_id: str
    branch_id: str
    branch_name: str = "Branch"
    branch_role: str = "explore_alternatives"
    branch_status: str = "active"
    child_thread_id: str | None = None
    parent_thread_id: str | None = None
    findings: list[dict[str, Any]] = Field(default_factory=list)
    user_id: str | None = None
    auto_promote: bool | None = None


class AgentMemoryCuratorEvaluateResponse(BaseModel):
    decision: dict[str, Any] = Field(default_factory=dict)


class AgentMemoryCuratorDecisionListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentDelegationPolicyResponse(BaseModel):
    enabled: bool = False
    enforce: bool = False
    max_parallel_runs: int = 1
    default_off_legacy_safe: bool = True


class AgentDelegationPlanRequest(BaseModel):
    message: str
    scene: str = "agent_delegation_console"
    available_tools: list[str] = Field(default_factory=list)


class AgentDelegationPlanResponse(BaseModel):
    policy: AgentDelegationPolicyResponse
    plan: dict[str, Any] = Field(default_factory=dict)


class AgentDelegationRunListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentModelRouterPolicyResponse(BaseModel):
    enabled: bool = False
    mode: str = "observe"
    default_model: str
    helper_model: str | None = None
    role_models: dict[str, str | None] = Field(default_factory=dict)


class AgentModelRouteRequest(BaseModel):
    role: str = "executor"
    selected_model: str | None = None
    task_text: str = ""
    tool_risk: str = "low"
    context_size: int = 0


class AgentModelRouteResponse(BaseModel):
    decision: dict[str, Any] = Field(default_factory=dict)


class AgentModelRouterDecisionListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentSelfRepairFailureListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentSelfRepairPromotePreviewRequest(BaseModel):
    failures: list[dict[str, Any]] = Field(default_factory=list)
    case_id_prefix: str = "agent_delegation"


class AgentSelfRepairPromotePreviewResponse(BaseModel):
    preview: dict[str, Any] = Field(default_factory=dict)


class AgentReviewQueueListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentReviewQueueDecisionResponse(BaseModel):
    item: dict[str, Any] = Field(default_factory=dict)


class AgentContextPolicyResponse(BaseModel):
    enabled: bool = False
    artifactize_long_observations: bool = False
    role_views_enabled: bool = False
    tokenizer_mode: str = "tokenizer_first"
    artifact_min_chars: int = 12000
    default_off_legacy_safe: bool = True


class AgentContextPreviewRequest(BaseModel):
    state: dict[str, Any] = Field(default_factory=dict)
    prompt_mode: str = "explore"
    role: str = "executor"
    assembled_context: str | None = None
    materialize_artifacts: bool = False


class AgentContextPreviewResponse(BaseModel):
    decision: dict[str, Any] = Field(default_factory=dict)


class AgentContextDecisionListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentContextArtifactListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentTaskLedgerPolicyResponse(BaseModel):
    enabled: bool = False
    artifact_synthesis_enabled: bool = False
    critic_gate_enabled: bool = False
    critic_gate_enforce: bool = False
    default_off_legacy_safe: bool = True


class AgentTaskLedgerPlanRequest(BaseModel):
    message: str = ""
    delegation_plan: dict[str, Any] = Field(default_factory=dict)


class AgentTaskLedgerPlanResponse(BaseModel):
    policy: AgentTaskLedgerPolicyResponse
    ledger: dict[str, Any] = Field(default_factory=dict)
    artifacts: list[dict[str, Any]] = Field(default_factory=list)
    critic_gate_result: dict[str, Any] | None = None
    synthesis_result: dict[str, Any] | None = None


class AgentTaskLedgerRunListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentArtifactListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentArtifactSynthesisRequest(BaseModel):
    artifacts: list[dict[str, Any]] = Field(default_factory=list)
    critic_gate_result: dict[str, Any] | None = None


class AgentArtifactSynthesisResponse(BaseModel):
    result: dict[str, Any] = Field(default_factory=dict)


class AgentCriticVerdictListResponse(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)
    count: int = 0
    trajectory_available: bool = False
    trajectory_error: str | None = None


class AgentCriticEvaluateRequest(BaseModel):
    ledger: dict[str, Any] = Field(default_factory=dict)
    artifacts: list[dict[str, Any]] = Field(default_factory=list)


class AgentCriticEvaluateResponse(BaseModel):
    result: dict[str, Any] = Field(default_factory=dict)


__all__ = [
    "AgentRolePolicyResponse",
    "AgentSkillSelectRequest",
    "AgentSkillSemanticCandidateResponse",
    "AgentSkillSelectionResponse",
    "AgentSkillSelectionEventResponse",
    "AgentSkillSelectionEventListResponse",
    "AgentSkillSelectionFeedbackRequest",
    "AgentSkillSelectionFeedbackResponse",
    "AgentFeedbackTrendResponse",
    "AgentSkillPreferenceRequest",
    "AgentSkillPreferenceResponse",
    "AgentSkillCatalogItemResponse",
    "AgentSkillCatalogResponse",
    "AgentContextExplainRequest",
    "AgentContextExplainResponse",
    "AgentContextEvidenceResponse",
    "AgentContextEvidenceListResponse",
    "AgentRoleDryRunRequest",
    "AgentRoleDryRunResponse",
    "AgentRoleDecisionListResponse",
    "AgentCapabilityResponse",
    "AgentCapabilityListResponse",
    "AgentToolRouteRequest",
    "AgentToolRouteResponse",
    "AgentToolRouteDecisionListResponse",
    "AgentMemoryCuratorPolicyResponse",
    "AgentMemoryCuratorEvaluateRequest",
    "AgentMemoryCuratorEvaluateResponse",
    "AgentMemoryCuratorDecisionListResponse",
    "AgentDelegationPolicyResponse",
    "AgentDelegationPlanRequest",
    "AgentDelegationPlanResponse",
    "AgentDelegationRunListResponse",
    "AgentModelRouterPolicyResponse",
    "AgentModelRouteRequest",
    "AgentModelRouteResponse",
    "AgentModelRouterDecisionListResponse",
    "AgentSelfRepairFailureListResponse",
    "AgentSelfRepairPromotePreviewRequest",
    "AgentSelfRepairPromotePreviewResponse",
    "AgentReviewQueueListResponse",
    "AgentReviewQueueDecisionResponse",
    "AgentContextPolicyResponse",
    "AgentContextPreviewRequest",
    "AgentContextPreviewResponse",
    "AgentContextDecisionListResponse",
    "AgentContextArtifactListResponse",
    "AgentTaskLedgerPolicyResponse",
    "AgentTaskLedgerPlanRequest",
    "AgentTaskLedgerPlanResponse",
    "AgentTaskLedgerRunListResponse",
    "AgentArtifactListResponse",
    "AgentArtifactSynthesisRequest",
    "AgentArtifactSynthesisResponse",
    "AgentCriticVerdictListResponse",
    "AgentCriticEvaluateRequest",
    "AgentCriticEvaluateResponse",
]

File: frontend-sdk/src/types/agent.ts
export interface FocusAgentModelOption {
  id: string;
  provider: string;
  provider_label: string;
  provider_logo_slug?: string | null;
  provider_logo_letter?: string | null;
  name: string;
  label: string;
  is_default: boolean;
  supports_thinking: boolean;
  default_thinking_enabled: boolean;
}

export interface FocusAgentModelsResponse {
  default_model: string;
  models: FocusAgentModelOption[];
}

export interface FocusAgentRolePolicyResponse {
  enabled: boolean;
  default_model: string;
  helper_model?: string | null;
  max_parallel_runs: number;
  roles: string[];
  role_models: Record<string, string | null>;
  fallback_order: string[];
}

export interface FocusAgentRoleDryRunRequest {
  message: string;
  scene?: string;
  skill_hints?: string[];
  available_tools?: string[];
}

export interface FocusAgentRoleDryRunResponse {
  policy: FocusAgentRolePolicyResponse;
  plan: Record<string, unknown>;
}

export interface FocusAgentSkillSelectRequest {
  message: string;
  skill_hints?: string[];
  semantic_enabled?: boolean | null;
  semantic_threshold?: number | null;
}

export interface FocusAgentSkillSemanticCandidate {
  skill_id: string;
  score: number;
  matched_terms: string[];
  auto_activate: boolean;
  rationale: string;
}

export interface FocusAgentSkillSelectionResponse {
  skill_ids: string[];
  stripped_message: string;
  prompt_mode?: string | null;
  selection_source: string;
  matched_triggers: string[];
  semantic_candidates: FocusAgentSkillSemanticCandidate[];
  confidence: number;
  rationale: string;
  semantic_enabled: boolean;
  semantic_threshold: number;
}

export interface FocusAgentSkillSelectionEvent {
  selection_id: string;
  created_at: string;
  message?: string | null;
  selection_source: string;
  explicit_hints: string[];
  matched_triggers: string[];
  semantic_candidates: FocusAgentSkillSemanticCandidate[];
  activated_skills: string[];
  confidence: number;
  rationale?: string | null;
  user_override?: string | null;
  feedback?: string | null;
}

export interface FocusAgentSkillSelectionListResponse {
  items: FocusAgentSkillSelectionEvent[];
  count: number;
}

export interface FocusAgentSkillCatalogItem {
  skill_id: string;
  description: string;
  when_to_use: string[];
  triggers: string[];
  recommended_tools: string[];
  path?: string | null;
  prompt_mode?: string | null;
  preference?: FocusAgentSkillPreferenceResponse | null;
}

export interface FocusAgentSkillCatalogResponse {
  items: FocusAgentSkillCatalogItem[];
  count: number;
}

export interface FocusAgentSkillSelectionFeedbackRequest {
  feedback: "useful" | "misfire" | "ignored" | string;
  note?: string | null;
}

export interface FocusAgentSkillPreferenceRequest {
  state: string;
  metadata?: Record<string, unknown>;
}

export interface FocusAgentSkillPreferenceResponse {
  preference_id: string;
  user_id: string;
  skill_id: string;
  state: string;
  metadata: Record<string, unknown>;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface FocusAgentRoleDecisionListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentCapability {
  name: string;
  description: string;
  toolset?: string | null;
  allowed_roles: string[];
  risk_level: string;
  side_effect: boolean;
  parallel_safe: boolean;
  cacheable: boolean;
  requires_network: boolean;
  requires_workspace_write: boolean;
  requires_approval: boolean;
  sensitive_args?: string[];
  redaction_policy?: string;
  provider_id?: string | null;
}

export interface FocusAgentCapabilityListResponse {
  items: FocusAgentCapability[];
  count: number;
}

export interface FocusAgentToolset {
  name: string;
  description: string;
  tools: string[];
  count: number;
  provider_ids: string[];
  risk_levels: string[];
  allowed_roles: string[];
  intent_policies: string[];
  requires_network: boolean;
  requires_workspace_write: boolean;
  side_effect: boolean;
  requires_approval: boolean;
}

export interface FocusAgentToolsetListResponse {
  items: FocusAgentToolset[];
  count: number;
}

export interface FocusAgentToolRouteRequest {
  role?: string;
  tool_policy?: string;
  available_tools?: string[];
  enforce?: boolean | null;
}

export interface FocusAgentToolRouteResponse {
  plan: Record<string, unknown>;
}

export interface FocusAgentToolRouteDecisionListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentMemoryCuratorPolicyResponse {
  enabled: boolean;
  auto_promote_on_merge: boolean;
  branch_local_only_until_merge: boolean;
  conflict_strategy: string;
}

export interface FocusAgentMemoryCuratorEvaluateRequest {
  root_thread_id: string;
  branch_id: string;
  branch_name?: string;
  branch_role?: string;
  branch_status?: string;
  child_thread_id?: string | null;
  parent_thread_id?: string | null;
  findings?: Array<Record<string, unknown>>;
  user_id?: string | null;
  auto_promote?: boolean | null;
}

export interface FocusAgentMemoryCuratorEvaluateResponse {
  decision: Record<string, unknown>;
}

export interface FocusAgentMemoryCuratorDecisionListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentDelegationPolicyResponse {
  enabled: boolean;
  enforce: boolean;
  max_parallel_runs: number;
  default_off_legacy_safe: boolean;
}

export interface FocusAgentDelegationPlanRequest {
  message: string;
  scene?: string;
  available_tools?: string[];
}

export interface FocusAgentDelegationPlanResponse {
  policy: FocusAgentDelegationPolicyResponse;
  plan: Record<string, unknown>;
}

export interface FocusAgentDelegationRunListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentModelRouterPolicyResponse {
  enabled: boolean;
  mode: string;
  default_model: string;
  helper_model?: string | null;
  role_models: Record<string, string | null>;
}

export interface FocusAgentModelRouteRequest {
  role?: string;
  selected_model?: string | null;
  task_text?: string;
  tool_risk?: string;
  context_size?: number;
}

export interface FocusAgentModelRouteResponse {
  decision: Record<string, unknown>;
}

export interface FocusAgentModelRouterDecisionListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentSelfRepairFailureListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentSelfRepairPromotePreviewRequest {
  failures?: Array<Record<string, unknown>>;
  case_id_prefix?: string;
}

export interface FocusAgentSelfRepairPromotePreviewResponse {
  preview: Record<string, unknown>;
}

export interface FocusAgentReviewQueueListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentReviewQueueDecisionResponse {
  item: Record<string, unknown>;
}

export interface FocusAgentContextPolicyResponse {
  enabled: boolean;
  artifactize_long_observations: boolean;
  role_views_enabled: boolean;
  tokenizer_mode: string;
  artifact_min_chars: number;
  default_off_legacy_safe: boolean;
}

export interface FocusAgentContextPreviewRequest {
  state?: Record<string, unknown>;
  prompt_mode?: string;
  role?: string;
  assembled_context?: string | null;
  materialize_artifacts?: boolean;
}

export interface FocusAgentContextPreviewResponse {
  decision: Record<string, unknown>;
}

export interface FocusAgentContextDecisionListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentContextArtifactListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentContextMemoryEvidence {
  evidence_id: string;
  thread_id?: string | null;
  turn_id?: string | null;
  created_at: string;
  selected_memories: Array<Record<string, unknown>>;
  excluded_memories: Array<Record<string, unknown>>;
  compaction_summary?: string | null;
  drift_report?: Record<string, unknown> | null;
  artifact_refs: Array<Record<string, unknown>>;
  token_counting_backend?: string | null;
  tokenizer_id?: string | null;
  estimated?: boolean | null;
  risk_flags: string[];
  metadata?: Record<string, unknown> | null;
}

export interface FocusAgentContextMemoryEvidenceListResponse {
  items: FocusAgentContextMemoryEvidence[];
  count: number;
}

export interface FocusAgentContextExplainRequest {
  thread_id?: string | null;
  turn_id?: string | null;
  message?: string | null;
}

export interface FocusAgentContextExplainResponse {
  evidence: FocusAgentContextMemoryEvidence;
  answerability?: Record<string, unknown> | null;
}

export interface FocusAgentMemoryUsageResponse {
  memory_id: string;
  usage: Array<Record<string, unknown>>;
  count: number;
}

export interface FocusAgentFeedbackTrendResponse {
  negative_feedback_count: number;
  merge_review_apply_success_rate?: number | null;
  merge_review_conflict_rate?: number | null;
  skill_low_confidence_rate?: number | null;
  skill_override_rate?: number | null;
  context_high_drift_count: number;
  notes_tasks_capture_count: number;
  top_failing_trajectory_samples: Array<Record<string, unknown>>;
  generated_at?: string | null;
}

export interface FocusAgentTaskLedgerPolicyResponse {
  enabled: boolean;
  artifact_synthesis_enabled: boolean;
  critic_gate_enabled: boolean;
  critic_gate_enforce: boolean;
  default_off_legacy_safe: boolean;
}

export interface FocusAgentTaskLedgerPlanRequest {
  message?: string;
  delegation_plan?: Record<string, unknown>;
}

export interface FocusAgentTaskLedgerPlanResponse {
  policy: FocusAgentTaskLedgerPolicyResponse;
  ledger: Record<string, unknown>;
  artifacts: Array<Record<string, unknown>>;
  critic_gate_result?: Record<string, unknown> | null;
  synthesis_result?: Record<string, unknown> | null;
}

export interface FocusAgentTaskLedgerRunListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentArtifactListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentArtifactSynthesisRequest {
  artifacts?: Array<Record<string, unknown>>;
  critic_gate_result?: Record<string, unknown> | null;
}

export interface FocusAgentArtifactSynthesisResponse {
  result: Record<string, unknown>;
}

export interface FocusAgentCriticVerdictListResponse {
  items: Array<Record<string, unknown>>;
  count: number;
  trajectory_available: boolean;
  trajectory_error?: string | null;
}

export interface FocusAgentCriticEvaluateRequest {
  ledger?: Record<string, unknown>;
  artifacts?: Array<Record<string, unknown>>;
}

export interface FocusAgentCriticEvaluateResponse {
  result: Record<string, unknown>;
}
